// @elastic/charts/dist/specs/specs_parser.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_react_1 = require("mobx-react");
var react_1 = require("react");
var SpecsSpecRootComponent = /** @class */ (function (_super) {
    __extends(SpecsSpecRootComponent, _super);
    function SpecsSpecRootComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SpecsSpecRootComponent.prototype.componentDidMount = function () {
        this.props.chartStore.specsInitialized.set(true);
        this.props.chartStore.computeChart();
    };
    SpecsSpecRootComponent.prototype.componentDidUpdate = function () {
        this.props.chartStore.specsInitialized.set(true);
        this.props.chartStore.computeChart();
    };
    SpecsSpecRootComponent.prototype.componentWillUnmount = function () {
        this.props.chartStore.chartInitialized.set(false);
    };
    SpecsSpecRootComponent.prototype.render = function () {
        return this.props.children || null;
    };
    return SpecsSpecRootComponent;
}(react_1.PureComponent));
exports.SpecsSpecRootComponent = SpecsSpecRootComponent;
exports.SpecsParser = mobx_react_1.inject('chartStore')(SpecsSpecRootComponent);
