// @elastic/charts/dist/chart_types/partition_chart/specs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __1 = require("../..");
var constants_1 = require("../../../specs/constants");
var spec_factory_1 = require("../../../state/spec_factory");
var config_1 = require("../layout/config/config");
var group_by_rollup_1 = require("../layout/utils/group_by_rollup");
var defaultProps = {
    chartType: __1.ChartTypes.Partition,
    specType: constants_1.SpecTypes.Series,
    config: config_1.config,
    valueAccessor: function (d) { return (typeof d === 'number' ? d : 0); },
    valueGetter: function (n) { return n[group_by_rollup_1.AGGREGATE_KEY]; },
    valueFormatter: function (d) { return String(d); },
    percentFormatter: config_1.percentFormatter,
    topGroove: 20,
    layers: [
        {
            groupByRollup: function (d, i) { return i; },
            nodeLabel: function (d) { return String(d); },
            showAccessor: function () { return true; },
            fillLabel: {},
        },
    ],
};
exports.Partition = spec_factory_1.getConnect()(spec_factory_1.specComponentFactory(defaultProps));
