// @kbn/ace/target/index.js
"use strict";
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.installXJsonMode = exports.XJsonMode = exports.addXJsonToRules = exports.XJsonHighlightRules = exports.ScriptHighlightRules = exports.ElasticsearchSqlHighlightRules = void 0;
var modes_1 = require("./ace/modes");
Object.defineProperty(exports, "ElasticsearchSqlHighlightRules", { enumerable: true, get: function () { return modes_1.ElasticsearchSqlHighlightRules; } });
Object.defineProperty(exports, "ScriptHighlightRules", { enumerable: true, get: function () { return modes_1.ScriptHighlightRules; } });
Object.defineProperty(exports, "XJsonHighlightRules", { enumerable: true, get: function () { return modes_1.XJsonHighlightRules; } });
Object.defineProperty(exports, "addXJsonToRules", { enumerable: true, get: function () { return modes_1.addXJsonToRules; } });
Object.defineProperty(exports, "XJsonMode", { enumerable: true, get: function () { return modes_1.XJsonMode; } });
Object.defineProperty(exports, "installXJsonMode", { enumerable: true, get: function () { return modes_1.installXJsonMode; } });
