// @elastic/charts/dist/chart_types/xy_chart/specs/axis.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_react_1 = require("mobx-react");
var react_1 = require("react");
var specs_1 = require("../utils/specs");
var ids_1 = require("../../../utils/ids");
var AxisSpec = /** @class */ (function (_super) {
    __extends(AxisSpec, _super);
    function AxisSpec() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AxisSpec.prototype.componentDidMount = function () {
        var _a = this.props, chartStore = _a.chartStore, children = _a.children, spec = __rest(_a, ["chartStore", "children"]);
        chartStore.addAxisSpec(__assign({}, spec));
    };
    AxisSpec.prototype.componentDidUpdate = function () {
        var _a = this.props, chartStore = _a.chartStore, children = _a.children, spec = __rest(_a, ["chartStore", "children"]);
        chartStore.addAxisSpec(__assign({}, spec));
    };
    AxisSpec.prototype.componentWillUnmount = function () {
        var id = this.props.id;
        this.props.chartStore.removeAxisSpec(id);
    };
    AxisSpec.prototype.render = function () {
        return null;
    };
    AxisSpec.defaultProps = {
        groupId: ids_1.getGroupId(specs_1.DEFAULT_GLOBAL_ID),
        hide: false,
        showOverlappingTicks: false,
        showOverlappingLabels: false,
        position: specs_1.Position.Left,
        tickSize: 10,
        tickPadding: 10,
        tickFormat: function (tick) { return "" + tick; },
        tickLabelRotation: 0,
    };
    return AxisSpec;
}(react_1.PureComponent));
exports.Axis = mobx_react_1.inject('chartStore')(AxisSpec);
