// @elastic/charts/dist/chart_types/xy_chart/specs/line_series.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_react_1 = require("mobx-react");
var react_1 = require("react");
var specs_1 = require("../utils/specs");
var ids_1 = require("../../../utils/ids");
var scales_1 = require("../../../utils/scales/scales");
var LineSeriesSpecComponent = /** @class */ (function (_super) {
    __extends(LineSeriesSpecComponent, _super);
    function LineSeriesSpecComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LineSeriesSpecComponent.prototype.componentDidMount = function () {
        var _a = this.props, chartStore = _a.chartStore, children = _a.children, config = __rest(_a, ["chartStore", "children"]);
        chartStore.addSeriesSpec(__assign({}, config));
    };
    LineSeriesSpecComponent.prototype.componentDidUpdate = function (prevProps) {
        var _a = this.props, chartStore = _a.chartStore, children = _a.children, config = __rest(_a, ["chartStore", "children"]);
        chartStore.addSeriesSpec(__assign({}, config));
        if (prevProps.id !== this.props.id) {
            chartStore.removeSeriesSpec(prevProps.id);
        }
    };
    LineSeriesSpecComponent.prototype.componentWillUnmount = function () {
        var _a = this.props, chartStore = _a.chartStore, id = _a.id;
        chartStore.removeSeriesSpec(id);
    };
    LineSeriesSpecComponent.prototype.render = function () {
        return null;
    };
    LineSeriesSpecComponent.defaultProps = {
        seriesType: 'line',
        groupId: ids_1.getGroupId(specs_1.DEFAULT_GLOBAL_ID),
        xScaleType: scales_1.ScaleType.Ordinal,
        yScaleType: scales_1.ScaleType.Linear,
        xAccessor: 'x',
        yAccessors: ['y'],
        yScaleToDataExtent: false,
        hideInLegend: false,
        histogramModeAlignment: specs_1.HistogramModeAlignments.Center,
    };
    return LineSeriesSpecComponent;
}(react_1.PureComponent));
exports.LineSeriesSpecComponent = LineSeriesSpecComponent;
exports.LineSeries = mobx_react_1.inject('chartStore')(LineSeriesSpecComponent);
